pallets/gear-debug/src/tests.rs

```diff
@@ -300,3 +282,227 @@ fn get_last_message_id() -> MessageId {
+                    i32.const 0xa9876
+                    i32.load
+                    i32.eqz             ;; must be zero by default
+                    br_if 0
+                    unreachable
+                )
+
+                ;; access page 11 (dynamic)
+                i32.const 0xb8765
+                i32.const 0x42
+                i32.store
+
+                ;; then free page 11
+                i32.const 11
+                call $free
+            )
+
+            (func $handle
+                (block
+                    ;; check page 1 data
+                    i32.const 0x10002
+                    i64.load
+                    i64.const 0x3038373635343332  ;; is "23456780", "0" because we change it in init
+                    i64.eq
+                    br_if 0
+                    unreachable
+                )
+                (block
+                    ;; check page 7 data
+                    i32.const 0x70001
+                    i32.load
+                    i32.const 0x52414547 ;; is "GEAR"
+                    i32.eq
+                    br_if 0
+                    unreachable
+                )
+                (block
+                    ;; check page 8 data
+                    ;; currently free + allocation must save page data,
+                    ;; but this behavior may change in future.
+                    i32.const 0x87654
+                    i32.load
+                    i32.const 0x42
+                    i32.eq
+                    br_if 0
+                    unreachable
+                )
+                (block
+                    ;; check page 9 data
+                    i32.const 0x98765
+                    i32.load
+                    i32.const 0x42
+                    i32.eq
+                    br_if 0
+                    unreachable
+                )
+                ;; change page 3 and 4
+                ;; because we store 0x00_00_00_42 then bits will be changed
+                ;; in 3th page only, so the 3th page only must be in storage.
+                i32.const 0x3fffd
+                i32.const 0x42
+                i32.store
+            )
+
+            (data $.rodata (i32.const 0x10000) "0123456789")
+            (data $.rodata (i32.const 0x70001) "GEAR TECH")
+        )
+    "#;
+
+    init_logger();
+    new_test_ext().execute_with(|| {
+        let code = parse_wat(wat);
+        let program_id = generate_program_id(&code);
+        let origin = Origin::signed(1);
+
+        PalletGear::<Test>::submit_program(
+            origin.clone(),
+            code.clone(),
+            b"salt".to_vec(),
+            Vec::new(),
+            100_000_000_u64,
+            0_u128,
+        )
+        .expect("Failed to submit program");
+
+        // Enable debug-mode
+        DebugMode::<Test>::put(true);
+
+        run_to_block(2, None);
+
+        Pallet::<Test>::do_snapshot();
+
+        let static_pages = WasmPageNumber(8);
+
+        let page1_addr = 0x10000;
```

Comment:
damn, provide this values at the start of the test and use them inside wat, please
as I remember, formatting raw string forks well

Reply:
Looks good for me, because
1) it's just a test
2) it's better to view values as is inside wasm

Reply:
nevermind